### PR TITLE
Fix: make the stencil test functions usable.

### PR DESCRIPTION
**Pull request by zzuegg**
Hey. 

Long time since my last change to the engine. Here we go. 
This fix makes the stencil functions usable. They where hard coded to 0 as reference and Integer.MaxValue as mask

This fix does not change the old behavior. If you do not change them manually they are still 0/Max

### CODE REVIEW

File: jme3-core/src/main/java/com/jme3/material/RenderState.java [hunk 1]
```diff
@@ -551,6 +559,10 @@ public void read(JmeImporter im) throws IOException {
         backStencilDepthPassOperation = ic.readEnum("backStencilDepthPassOperation", StencilOperation.class, StencilOperation.Keep);
         frontStencilFunction = ic.readEnum("frontStencilFunction", TestFunction.class, TestFunction.Always);
         backStencilFunction = ic.readEnum("backStencilFunction", TestFunction.class, TestFunction.Always);
+        frontStencilReference=ic.readInt("frontStencilReference",0);
+        backStencilReference=ic.readInt("backStencilReference",0);
+        frontStencilMask=ic.readInt("frontStencilMask",Integer.MAX_VALUE);
+        backStencilMask=ic.readInt("backStencilMask",Integer.MAX_VALUE);
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/material/RenderState.java:565]: Please fix the formatting issue. (Add space before and after "=" and add space after ",")

File: jme3-core/src/main/java/com/jme3/material/RenderState.java [hunk 2]
```diff
@@ -508,6 +512,10 @@ public void write(JmeExporter ex) throws IOException {
         oc.write(backStencilDepthPassOperation, "backStencilDepthPassOperation", StencilOperation.Keep);
         oc.write(frontStencilFunction, "frontStencilFunction", TestFunction.Always);
         oc.write(backStencilFunction, "backStencilFunction", TestFunction.Always);
+        oc.write(frontStencilReference,"frontStencilReference",0);
+        oc.write(backStencilReference,"backStencilReference",0);
+        oc.write(frontStencilMask,"frontStencilMask",Integer.MAX_VALUE);
+        oc.write(backStencilMask,"backStencilMask",Integer.MAX_VALUE);
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/material/RenderState.java:518]: Please fix the formatting issue. (add space after ",")

File: jme3-core/src/main/java/com/jme3/material/RenderState.java [hunk 3]
```diff
@@ -480,6 +480,10 @@ public enum StencilOperation {
     StencilOperation backStencilDepthPassOperation = StencilOperation.Keep;
     TestFunction frontStencilFunction = TestFunction.Always;
     TestFunction backStencilFunction = TestFunction.Always;
+    int frontStencilReference=0;
+    int backStencilReference=0;
+    int frontStencilMask=Integer.MAX_VALUE;
+    int backStencilMask=Integer.MAX_VALUE;
```
Comment (Ali-RS) [inline jme3-core/src/main/java/com/jme3/material/RenderState.java:486]: Please fix the formatting issue. (Add space before and after "=")

File: jme3-core/src/main/java/com/jme3/material/RenderState.java [hunk 4]
```diff
@@ -695,6 +707,18 @@ public boolean equals(Object o) {
...
+            }
+            if(backStencilMask != rs.backStencilMask){
+                return false;
+            }
+            if(frontStencilReference != rs.frontStencilReference){
+                return false;
+            }
+            if(backStencilReference != rs.backStencilReference){
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (zzuegg) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java:721]: My added if() statements have the same formatting as the once already there. How to resolve? Should i reformat the already existing or leave mine as they are. (Looks strange if half the if's are formatted differently)

[the remark this answers is not in the document]
Comment (pspeed42) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java:721]: Personally, I believe you did the right thing by adopting the style that was already in the rest of the file... even if it's not a style that I like.  Others may have a different opinion but I'm ok with it.

File: jme3-core/src/main/java/com/jme3/material/RenderState.java (hunk 3 above)
Comment (zzuegg) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java]: done

File: jme3-core/src/main/java/com/jme3/material/RenderState.java (hunk 2 above)
Comment (zzuegg) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java]: done

File: jme3-core/src/main/java/com/jme3/material/RenderState.java (hunk 1 above)
Comment (zzuegg) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java]: done

File: jme3-core/src/main/java/com/jme3/material/RenderState.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ali-RS) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java:721]: Agree with Paul, but in this case, the existing ones are already formatted.

Existing ones:

```
            if (frontStencilStencilFailOperation != rs.frontStencilStencilFailOperation) {
                return false;
            }
            if (frontStencilDepthFailOperation != rs.frontStencilDepthFailOperation) {
                return false;
            }
            if (frontStencilDepthPassOperation != rs.frontStencilDepthPassOperation) {
                return false;
            }
            if (backStencilStencilFailOperation != rs.backStencilStencilFailOperation) {
                return false;
            }
            if (backStencilDepthFailOperation != rs.backStencilDepthFailOperation) {
                return false;
            }
            if (backStencilDepthPassOperation != rs.backStencilDepthPassOperation) {
                return false;
            }
            if (frontStencilFunction != rs.frontStencilFunction) {
                return false;
            }
            if (backStencilFunction != rs.backStencilFunction) {
                return false;
            }
```
the ones you added:

```
            if(frontStencilMask != rs.frontStencilMask){
                return false;
            }
            if(backStencilMask != rs.backStencilMask){
                return false;
            }
            if(frontStencilReference != rs.frontStencilReference){
                return false;
            }
            if(backStencilReference != rs.backStencilReference){
                return false;
            }
```

[the remark this answers is not in the document]
Comment (zzuegg) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java:721]: missed the whitespace before and after the brackets. sorry goung to fix

File: jme3-core/src/main/java/com/jme3/material/RenderState.java (hunk 1 above)
Comment (Ali-RS) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java]: thanks

File: jme3-core/src/main/java/com/jme3/material/RenderState.java (hunk 2 above)
Comment (Ali-RS) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java]: thanks

File: jme3-core/src/main/java/com/jme3/material/RenderState.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zzuegg) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java]: done

[the remark this answers is not in the document]
Comment (Ali-RS) [reply jme3-core/src/main/java/com/jme3/material/RenderState.java]: Thank you

**Comment by Ali-RS** [comment]
Just a note, when merging please use the "Squash and merge" option to merge, so all commits will be combined into a single commit in the master branch.

**Comment by zzuegg** [comment]
Should i merge, or let someone of the team do it?

**Comment by Ali-RS** [comment]
>or let someone of the team do it?

well, it might take forever...

I think this is a relatively simple change and does not require much input from other members. But I would suggest letting it be open for a few days for other folks to have a chance to take a look and if there was no substantial discussion you can self-integrate it.

**Comment by Ali-RS** [comment]
I will appreciate it if you also update the copyright year in the modified classes.

**Review by riccardobl — APPROVED** [review]

**Comment by stephengold** [comment]
I notice that both stencil masks have 0x7fffffff as their default values. Was this intentional? Naively, I would've expected 0xffffffff ...

**Comment by zzuegg** [comment]
To be honest i copied over the values used prior to my changes without thinking about the actual bits. Nice catch. 

In this case (and i have rechecked the gl specs now) there are only images formats specified with stencil width of 8 bits.  So we have "no mask" by default.

**Pull request merged**